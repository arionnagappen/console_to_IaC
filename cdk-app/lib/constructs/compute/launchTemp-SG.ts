import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';

export interface SgProps {
  vpc: ec2.IVpc
}

export class AppSecurityGroup extends Construct {
  public readonly securityGroup: ec2.SecurityGroup;

  constructor(scope: Construct, id: string, props: SgProps) {
    super(scope, id);

    this.securityGroup = new ec2.SecurityGroup(this, 'AppServerSG', {
      vpc: props.vpc,
      description: 'App Server Security Group',  
      allowAllOutbound: false
    });

    
    // Only allow inbound traffic on port 80
    this.securityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80)
    )

    // Only allow outbound traffic on port 80
    this.securityGroup.addEgressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
    )

  }
}