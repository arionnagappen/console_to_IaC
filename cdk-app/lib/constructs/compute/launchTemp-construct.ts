import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { ISecurityGroup } from "aws-cdk-lib/aws-ec2";

export interface LaunchTempProps {
  launchTempName: string;
  imageId: string;
  instanceType: string;
  securityGroupId: ISecurityGroup["securityGroupId"];
}

export class MyLaunchTemp extends Construct {
  public readonly launchTemp: ec2.CfnLaunchTemplate

  constructor(scope: Construct, id: string, props: LaunchTempProps) {
    super(scope, id)

    this.launchTemp = new ec2.CfnLaunchTemplate(this, 'AppLaunchTemplate', {
      launchTemplateName: props.launchTempName,
      launchTemplateData: {
        imageId: props.imageId, 
        instanceType: props.instanceType,
        securityGroupIds: [props.securityGroupId],
      }
    });
  };
};