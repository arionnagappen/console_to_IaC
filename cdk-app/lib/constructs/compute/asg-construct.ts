import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { MyLaunchTemp } from './launchTemp-construct';
import { CfnAutoScalingGroup } from 'aws-cdk-lib/aws-autoscaling';
import { ISubnet } from 'aws-cdk-lib/aws-ec2';

export interface AsgProps {
  vpc: ec2.IVpc;
  launchTemplate: ec2.CfnLaunchTemplate;
  subnetType?: ec2.SubnetType;
  minSize?: number;
  maxSize?: number;
  desiredCapacity?: number;
}

export class MyAutoScalingGroup extends Construct {
  public readonly asg: CfnAutoScalingGroup;

  constructor(scope: Construct, id: string, props: AsgProps) {
    super(scope, id);

    const subnets = props.vpc.selectSubnets({
      subnetType: props.subnetType?? ec2.SubnetType.PUBLIC
    })

    this.asg = new CfnAutoScalingGroup(this, 'AppServerASG', {
      minSize: `${props.minSize ?? 1}`,
      maxSize: `${props.maxSize ?? 3}`,
      desiredCapacity: `${props.desiredCapacity ?? 2}`,
      vpcZoneIdentifier: subnets.subnetIds,
      launchTemplate: {
        launchTemplateId: props.launchTemplate.ref,
        version: props.launchTemplate.attrLatestVersionNumber
      }
    });
  }
}