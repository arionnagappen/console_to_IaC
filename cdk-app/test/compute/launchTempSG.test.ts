import * as cdk from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { AppSecurityGroup } from '../../lib/constructs/compute/launchTemp-SG';

// Assertion Test
test('App Server Security Group has correct rules', () => {
  // Instantiates the App Construct
  const app = new cdk.App();

  // Instantiates the Stack Construct & uses the App Construct as a parameter
  const stack = new cdk.Stack(app, 'TestStack');

  // Adds a Vpc
  const vpc = new ec2.Vpc(stack, 'TestVpc', {
    maxAzs: 1,
  });

  new AppSecurityGroup(stack, 'TestSG', {vpc});

  const template = Template.fromStack(stack);

  // Check that only 1 SG is creates
  template.resourceCountIs('AWS::EC2::SecurityGroup', 1);

  // Ingress Rule on Port 80
  template.hasResourceProperties('AWS::EC2::SecurityGroup', {
    SecurityGroupIngress: Match.arrayWith([
      Match.objectLike ({
        IpProtocol: 'tcp',
        FromPort: 80,
        ToPort: 80,
        CidrIp: '0.0.0.0/0',
      })
    ])
  });


  // Egress Rule on Port 80
  template.hasResourceProperties('AWS::EC2::SecurityGroup', {
    SecurityGroupEgress: Match.arrayWith([
      Match.objectLike ({
        IpProtocol: 'tcp',
        FromPort: 80,
        ToPort: 80,
        CidrIp: '0.0.0.0/0',
      })
    ])
  });
});

// Snapshot Test
test('App Security Group matches snapshot', ()=>{
  // Instantiates the App Construct
  const app = new cdk.App();

  // Instantiates the Stack Construct & uses the App Construct as a parameter
  const stack = new cdk.Stack(app, 'TestStack');

  // Adds a Vpc
  const vpc = new ec2.Vpc(stack, 'TestVpc', {maxAzs: 1,});

  new AppSecurityGroup(stack, 'TestSG', {vpc});

  const template = Template.fromStack(stack).toJSON();

  expect(template).toMatchSnapshot();
});