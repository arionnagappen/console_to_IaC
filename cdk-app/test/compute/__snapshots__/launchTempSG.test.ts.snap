// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Security Group matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestSGAppServerSG063DFCD3": {
      "Properties": {
        "GroupDescription": "App Server Security Group",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestVpcE77CE678": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcIGW9DD53F70": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcPrivateSubnet1DefaultRoute32E7B814": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "TestVpcPublicSubnet1NATGatewayA323E3EC",
        },
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet1RouteTable469B0105",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPrivateSubnet1RouteTable469B0105": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPrivateSubnet1RouteTableAssociationFFD4DFF7": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet1RouteTable469B0105",
        },
        "SubnetId": {
          "Ref": "TestVpcPrivateSubnet1SubnetCC65D771",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPrivateSubnet1SubnetCC65D771": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/17",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcPublicSubnet1DefaultRoute6C0F0315": {
      "DependsOn": [
        "TestVpcVPCGWF1827B84",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet1RouteTable4CBFF871",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPublicSubnet1EIP4884338C": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "TestVpcPublicSubnet1NATGatewayA323E3EC": {
      "DependsOn": [
        "TestVpcPublicSubnet1DefaultRoute6C0F0315",
        "TestVpcPublicSubnet1RouteTableAssociation7D1DECD9",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "TestVpcPublicSubnet1EIP4884338C",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet1SubnetA7DB1EDF",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "TestVpcPublicSubnet1RouteTable4CBFF871": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPublicSubnet1RouteTableAssociation7D1DECD9": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet1RouteTable4CBFF871",
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet1SubnetA7DB1EDF",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPublicSubnet1SubnetA7DB1EDF": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/17",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcVPCGWF1827B84": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
